include/vcpkg/base/expected.h

```diff
@@ -105,191 +26,332 @@ namespace vcpkg
+            if (value_is_error)
             {
-                return {f(*m_t.get()), expected_left_tag};
+                return {m_error, expected_right_tag};
             }
             else
             {
-                return ExpectedT<map_t<F>, S>{m_s};
+                return {f(*m_t.get()), expected_left_tag};
             }
         }
 
         template<class F>
-        using move_map_t =
-            decltype(std::declval<F&>()(std::move(*std::declval<typename ExpectedHolder<T>::pointer>())));
+        ExpectedT<decltype(std::declval<F&>()(std::declval<T>())), Error> map(F f) &&
+        {
+            if (value_is_error)
+            {
+                return {std::move(m_error), expected_right_tag};
+            }
+            else
+            {
+                return {f(std::move(*m_t.get())), expected_left_tag};
+            }
+        }
 
+        // map_error(F): returns an Expected<T, result_of_calling_F>
+        //
+        // If *this holds a value, returns an expected holding the same value.
+        // Otherwise, returns an expected containing f(error())
         template<class F>
-        ExpectedT<move_map_t<F>, S> map(F f) &&
+        ExpectedT<T, decltype(std::declval<F&>()(std::declval<const Error&>()))> map_error(F f) const&
         {
-            if (has_value())
+            if (value_is_error)
             {
-                return {f(std::move(*m_t.get())), expected_left_tag};
+                return {f(m_error), expected_right_tag};
             }
             else
             {
-                return ExpectedT<move_map_t<F>, S>{std::move(m_s)};
+                return {*m_t.get(), expected_left_tag};
             }
         }
 
         template<class F>
-        ExpectedT& replace_error(F&& specific_error_generator)
+        ExpectedT<T, decltype(std::declval<F&>()(std::declval<Error>()))> map_error(F f) &&
         {
-            if (m_s.has_error())
+            if (value_is_error)
             {
-                m_s.error() = std::forward<F>(specific_error_generator)();
+                return {f(std::move(m_error)), expected_right_tag};
+            }
+            else
+            {
+                return {std::move(*m_t.get()), expected_left_tag};
             }
-
-            return *this;
         }
 
+        // then: f(T, Args...)
+        // If *this contains a value, returns INVOKE(f, *get(), forward(args)...)
+        // Otherwise, returns error() put into the same type.
+    private:
+        template<class Test>
+        struct IsThenCompatibleExpected : std::false_type
+        {
+        };
+
+        template<class OtherTy>
+        struct IsThenCompatibleExpected<ExpectedT<OtherTy, Error>> : std::true_type
+        {
+        };
+
+    public:
         template<class F, class... Args>
-        std::invoke_result_t<F, const_ref_type, Args&&...> then(F f, Args&&... args) const&
+        std::invoke_result_t<F, const T&, Args...> then(F f, Args&&... args) const&
```

Comment:
Why is `Args&&` not required?

Reply:
Because both `Args` and `Args&&` name rvalues.
con
That is, given:

```
int returns_prvalue();
int&& returns_xvalue();
int& returns_lvalue();

template<class Arg> void consume(Arg&&);
```

`consume` can't tell the difference between `consume(returns_prvalue())` and `consume(returns_xvalue())`, since forwarding references only sense lvalue-or-rvalueness. (In both cases, `Arg` would deduce to `int`, giving `void consume<int>(int&&)`. (and `consume(returns_lvalue())` would deduce `int&` giving `void consume<int&>(int&)`))

Convention when implementing perfect forwarders like this is to not add the meaningless &&s when passing types through, for example http://eel.is/c++draft/func.invoke

Reply:
See also http://eel.is/c++draft/fig:basic.lval

<img width="326" alt="image" src="https://user-images.githubusercontent.com/1544943/170575873-12499911-7ffa-4870-8c21-31f219b9c1cb.png">